pkg/block/metadata/deletionmark.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package metadata
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"path"
+
+	"github.com/go-kit/kit/log"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+const (
+	// DeletionMarkFilename is the known json filename to store details about when block is marked for deletion.
+	DeletionMarkFilename = "deletion-mark.json"
+
+	// DeletionMarkVersion1 is the version of deletion-mark file supported by Thanos.
+	DeletionMarkVersion1 = 1
+)
+
+// ErrorDeletionMarkNotFound is the error when deletion-mark.json file is not found.
+var ErrorDeletionMarkNotFound = errors.New("deletion-mark.json not found")
+
+// DeletionMark stores block id and when block was marked for deletion.
+type DeletionMark struct {
+	// ID of the tsdb block.
+	ID ulid.ULID `json:"id"`
+
+	// DeletionTime is a unix timestamp of when the block was marked to be deleted.
+	DeletionTime int64 `json:"deletion_time"`
+
+	// Version of the file.
+	Version int `json:"version"`
+}
+
+// ReadDeletionMark reads the given deletion mark file from <dir>/deletion-mark.json in bucket.
+func ReadDeletionMark(ctx context.Context, bkt objstore.BucketReader, logger log.Logger, dir string) (*DeletionMark, error) {
+	deletionMarkFile := path.Join(dir, DeletionMarkFilename)
+
+	r, err := bkt.Get(ctx, deletionMarkFile)
+	if err != nil {
+		if bkt.IsObjNotFoundErr(err) {
+			return nil, ErrorDeletionMarkNotFound
+		}
+		return nil, errors.Wrapf(err, "get file: %v", deletionMarkFile)
+	}
+
+	defer runutil.CloseWithLogOnErr(logger, r, "close bkt deletion-mark reader")
+
+	metaContent, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, errors.Wrapf(err, "read file: %v", deletionMarkFile)
+	}
+
+	deletionMark := DeletionMark{}
+	if err := json.Unmarshal(metaContent, &deletionMark); err != nil {
+		return nil, errors.Wrapf(err, "unmarshal file: %v", deletionMarkFile)
```

Comment:
I think unmarshal error should be treated as file being uploaded?

Reply:
(we do same for meta.json)

Reply:
A quick check through the codebase for `json.Unmarshal` for meta, I see that the error returned is for `unmarshal meta.json for block` and not about file is being uploaded. I am happy to change the error string for deletion mark but please note we are using `unmarshal file` only.

Reply:
instances of where meta is downloaded is `loadMeta` function in `fetcher.go` and `DownloadMeta` in `block.go`

Reply:
I have added a custom implementation of error `ErrorUnmarshalDeletionMark` for handling partial deletion-mark.json files now.